core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -537,14 +533,14 @@ class GroupCoordinator(val brokerId: Int,
           // TODO: cut the socket connection to the client
         }
 
+        // TODO KAFKA-2720: only remove group in the background thread
         if (group.isEmpty) {
-          group.transitionTo(Dead)
+          groupManager.removeGroup(group)
```

Comment:
Ah, yeah, that makes sense. I think I never got the reason for the dead check before.